component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/service/LocalCacheService.java

```diff
@@ -17,71 +17,285 @@
+    private transient Supplier<ScheduledExecutorService> threadServiceGetter;
+
+    public LocalCacheService(final String plugin, final Supplier<Long> timer,
+            final Supplier<ScheduledExecutorService> threadServiceGetter) {
+        this.plugin = plugin;
+        this.timer = timer;
+        this.threadServiceGetter = threadServiceGetter;
+    }
+
+    /**
+     * Evict an object.
+     * 
+     * @param key key to evict.
+     */
     @Override
     public void evict(final String key) {
-        cache.remove(internalKey(key));
+        final String realKey = internalKey(key);
+
+        // use compute to be able to call release.
+        cache.compute(realKey, (String oldKey, ElementImpl oldElement) -> {
+            if (oldElement != null && oldElement.canBeEvict()) {
+                // ok to evict, so do release.
+                oldElement.release();
+                return null;
+            }
+            return oldElement;
+        });
     }
 
     @Override
-    public <T> void evictIfValue(final String key, final T expected) {
-        cache.remove(internalKey(key), new Element(expected, 0));
+    public void evictIfValue(final String key, final Object expected) {
+        final String realKey = internalKey(key);
+
+        // use compute to be able to call release.
+        cache.compute(realKey, (String oldKey, ElementImpl oldElement) -> {
+            if (oldElement != null && (Objects.equals(oldElement.getValue(), expected) || oldElement.canBeEvict())) {
+                // ok to evit, so do release.
+                oldElement.release();
+                return null;
+            }
+            return oldElement;
+        });
+
     }
 
     @Override
-    public <T> T computeIfAbsent(final String key, final long timeoutMs, final Supplier<T> value) {
-        final String internalKey = internalKey(key);
-        final Element element = cache
-                .compute(internalKey,
-                        (k, e) -> e == null || e.isExpired()
-                                ? new Element(value.get(), System.currentTimeMillis() + timeoutMs)
-                                : e);
-        if (element == null) {
-            return null;
+    public <T> T computeIfAbsent(final Class<T> expectedClass, final String key,
+            final Predicate<Element> toRemove, final long timeoutMs,
+            final Supplier<T> value) {
+
+        final Integer maxSize = this.getConfigValue(CacheConfiguration::getDefaultMaxSize, -1);
+        if (maxSize > 0 && this.cache.size() >= maxSize) {
+            this.clean(); // clean before add one element.
+            if (this.cache.size() >= maxSize) {
+                synchronized (this.cache) {
+                    while (this.cache.size() >= maxSize) {
+                        final String keyToRemove = this.cache.keySet().iterator().next();
+                        this.cache.remove(keyToRemove);
+                    }
+                }
+            }
```

Comment:
If maxSize is reached, you first do a clean() which removes all keys that can be removed but which can be limited by the MaxDeletionPerEvictionRun (which has sense) ... but after (if still over the maxsize allowed) you loop unconditionnaly on all keys and remove them one by one ?
Not sure about this process... Can we think a task that will run and do this ? 
(Just a proposal to be studied)
Below, you create a task for the eviction for each key which has a timeout defined...
Why not having only one task which will do the maintenance? 
ie : eviction of expired keys and check the cache size

Reply:
- The clean is inside the 'adding' function. So, if it oversize maxSize; it should be small oversizing. So, if on line 113, it's still oversize maxSize; then, there is a strong probability that it's because items have either a valid date either a method tell we can keep object. So, the one by one eviction until we reach MaxSize seems to me a good strategy.

think a task : You mean a task that run at scheduled time (risk is to remove last element) or put the loop in separated thread ? Second option will allow to optimize insertions if creation of new task is quicker than clean method.

a task for each key allow to remove old object without scanning all cache map.

Reply:
hum, not sure about the strategy but as commented before, we now can stick to this one and advice a bit later.